src/graph_notebook/magics/metadata.py

```diff
@@ -0,0 +1,141 @@
+"""
+
+import json
+import sys
+import re
+from typing import List
+from requests import Response
+from graph_notebook.visualization.template_retriever import retrieve_template
+pre_container_template = retrieve_template("pre_container.html")
+
+
+# class for individual metadata metric
+class Metric(object):
+    def __init__(self, name: str, friendly_name: str, default_value: any = "N/A"):
+        self.name = name    # used as key metadata.metrics
+        self.friendly_name = friendly_name
+        self.value = default_value
+
+    def set_value(self, value: any):
+        self.value = value
+
+
+# class for complete metadata object to output
+class Metadata(object):
+    def __init__(self):
+        self.metrics = {}
+
+    def insert_metric(self, new_metric: Metric):
+        self.metrics[new_metric.name] = new_metric
+
+    def bulk_insert_metrics(self, new_metrics: List[Metric]):
+        for metric in new_metrics:
+            self.insert_metric(metric)
+
+    def set_metric_value(self, metric_name, value):
+        self.metrics[metric_name].set_value(value)
+
+    def set_request_metrics(self, res: Response):
+        self.set_metric_value('request_time', 1000 * res.elapsed.total_seconds())
+        self.set_metric_value('status', res.status_code)
+        self.set_metric_value('status_ok', res.ok)
+        self.set_metric_value('resp_size', sys.getsizeof(res.content))
+
+    def set_gremlin_profile_metrics(self, profile: str):
+        # Query execution time in Neptune DB
+        querytime_regex = re.search(r'Query Execution: (.*?)\n', profile)
+        if querytime_regex:
+            self.set_metric_value('query_time', float(querytime_regex.group(1)))
+        # # of predicates
+        predicates_regex = re.search(r'# of predicates: (.*?)\n', profile)
+        if predicates_regex:
+            self.set_metric_value('predicates', int(predicates_regex.group(1)))
+        # Count of results
+        count_regex = re.search(r'Count: (.*?)\n', profile)
+        if count_regex:
+            self.set_metric_value('results', int(count_regex.group(1)))
+        # Serialization
+        serialization_regex = re.search(r'Serialization: (.*?)\n', profile)
+        if serialization_regex:
+            self.set_metric_value('seri_time', float(serialization_regex.group(1)))
+        # Size of results returned
+        results_size_regex = re.search(r'Response size \(bytes\): (.*?)\n', profile)
+        if results_size_regex:
+            self.set_metric_value('results_size', int(results_size_regex.group(1)))
+
+    def to_dict(self):
+        metadata_dict = {}
+        for metric in self.metrics.values():
+            metadata_dict[metric.friendly_name] = metric.value
+        return metadata_dict
+
+    def to_json(self):
+        return json.dumps(self.to_dict(), indent=4)
+
+    def to_html(self):
+        return pre_container_template.render(content=self.to_json())
+
+
+def create_gremlin_metadata_obj(q_mode: str) -> Metadata:
```

Comment:
Let's maintain this pattern for the gremlin profile metadata as well

Reply:
Done